Use consolidated writes/reads by default

Upstream is failing since https://github.com/pydata/xarray/pull/5252 went in, which enables consolidated writes/reads by default.

See e.g. https://github.com/pystatgen/sgkit/runs/2927778139?check_suite_focus=true

```
E               RuntimeWarning: Failed to open Zarr store with consolidated metadata, falling back to try reading non-consolidated metadata. This is typically much slower for opening a dataset. To silence this warning, consider:
E               1. Consolidating metadata in this existing store with zarr.consolidate_metadata().
E               2. Explicitly setting consolidated=False, to avoid trying to read consolidate metadata, or
E               3. Explicitly setting consolidated=True, to raise an error in this case instead of falling back to try reading non-consolidated metadata.

/opt/hostedtoolcache/Python/3.8.10/x64/lib/python3.8/site-packages/xarray/backends/zarr.py:371: RuntimeWarning
=========================== short test summary info ============================
FAILED sgkit/tests/test_regenie.py::test_regenie__glow_comparison - RuntimeWa...
FAILED sgkit/tests/test_vcfzarr_reader.py::test_vcfzarr_to_zarr[None-True-False-True]
FAILED sgkit/tests/test_vcfzarr_reader.py::test_vcfzarr_to_zarr[None-True-True-False]
FAILED sgkit/tests/io/bgen/test_bgen_reader.py::test_rechunk_bgen__target_chunks[target_chunks0]
FAILED sgkit/tests/io/bgen/test_bgen_reader.py::test_rechunk_bgen__target_chunks[target_chunks1]
FAILED sgkit/tests/io/bgen/test_bgen_reader.py::test_rechunk_bgen__target_chunks[target_chunks2]
FAILED sgkit/tests/io/bgen/test_bgen_reader.py::test_rechunk_bgen__target_chunks[target_chunks3]
FAILED sgkit/tests/io/bgen/test_bgen_reader.py::test_rechunk_from_zarr__self_consistent
FAILED sgkit/tests/io/bgen/test_bgen_reader.py::test_rechunk_bgen__probability_encoding[uint8]
FAILED sgkit/tests/io/bgen/test_bgen_reader.py::test_rechunk_bgen__probability_encoding[uint16]
FAILED sgkit/tests/io/bgen/test_bgen_reader.py::test_rechunk_bgen__variable_packing
FAILED sgkit/tests/io/bgen/test_bgen_reader.py::test_unpack_variables__invalid_gp_dims
FAILED sgkit/tests/io/bgen/test_bgen_reader.py::test_bgen_to_zarr[None] - Run...
FAILED sgkit/tests/io/bgen/test_bgen_reader.py::test_bgen_to_zarr[region1] - ...
FAILED sgkit/tests/io/bgen/test_bgen_reader.py::test_bgen_to_zarr[region2] - ...
FAILED sgkit/tests/io/vcf/test_vcf_reader.py::test_vcf_to_zarr__multiple_partitioned_by_size[True]
FAILED sgkit/tests/io/vcf/test_vcf_reader.py::test_vcf_to_zarr__multiple_partitioned_by_size[False]
```

--- a/docs/user_guide.rst
+++ b/docs/user_guide.rst
@@ -63,6 +63,9 @@ Similarly, the function :func:`sgkit.save_dataset` can be used to save the datas
 This can be used to save a newly-loaded dataset, or a dataset with new variables that are the
 result of running genetic methods on the data, so it can be loaded again later.
 
+By default, Zarr datasets created by sgkit (and Xarray) have `consolidated metadata <http://xarray.pydata.org/en/stable/user-guide/io.html#consolidated-metadata>`_,
+which makes opening the dataset faster.
+
 BGEN
 ----
 

--- a/sgkit/io/bgen/bgen_reader.py
+++ b/sgkit/io/bgen/bgen_reader.py
@@ -518,6 +518,8 @@ def rechunk_bgen(
         )
         rechunked.execute()
 
+    zarr.consolidate_metadata(output)
+
     ds: Dataset = xr.open_zarr(output, concat_characters=False)  # type: ignore[no-untyped-call]
     if pack:
         ds = unpack_variables(ds)

--- a/sgkit/io/utils.py
+++ b/sgkit/io/utils.py
@@ -74,14 +74,10 @@ def zarrs_to_dataset(
 
     datasets = []
     for path in urls:
-        try:
-            dataset = xr.open_zarr(  # type: ignore[no-untyped-call]
-                fsspec.get_mapper(path, **storage_options), concat_characters=False
-            )
-            datasets.append(dataset)
-        except zarr.errors.GroupNotFoundError:
-            # ignore empty directory, occurs if no variants in region
-            pass
+        dataset = xr.open_zarr(  # type: ignore[no-untyped-call]
+            fsspec.get_mapper(path, **storage_options), concat_characters=False
+        )
+        datasets.append(dataset)
 
     # Combine the datasets into one
     ds = xr.concat(datasets, dim="variants", data_vars="minimal")

--- a/sgkit/io/vcf/utils.py
+++ b/sgkit/io/vcf/utils.py
@@ -19,16 +19,23 @@ def ceildiv(a: int, b: int) -> int:
     return -(-a // b)
 
 
-# https://dev.to/orenovadia/solution-chunked-iterator-python-riddle-3ple
+# Based on https://dev.to/orenovadia/solution-chunked-iterator-python-riddle-3ple
 def chunks(iterator: Iterator[T], n: int) -> Iterator[Iterator[T]]:
     """
     Convert an iterator into an iterator of iterators, where the inner iterators
     each return `n` items, except the last, which may return fewer.
+
+    For the special case of an empty iterator, an iterator of an empty iterator is
+    returned.
     """
 
+    empty_iterator = True
     for first in iterator:  # take one item out (exits loop if `iterator` is empty)
+        empty_iterator = False
         rest_of_chunk = itertools.islice(iterator, 0, n - 1)
         yield itertools.chain([first], rest_of_chunk)  # concatenate the first item back
+    if empty_iterator:
+        yield iter([])
 
 
 def get_file_length(

--- a/sgkit/io/vcf/vcf_reader.py
+++ b/sgkit/io/vcf/vcf_reader.py
@@ -385,8 +385,9 @@ def vcf_to_zarr_sequential(
         for variants_chunk in chunks(region_filter(variants, region), chunk_length):
 
             variant_ids = []
-            variant_allele = []
+            variant_alleles = []
 
+            i = -1  # initialize in case of empty variants_chunk
             for i, variant in enumerate(variants_chunk):
                 variant_id = variant.ID if variant.ID is not None else "."
                 variant_ids.append(variant_id)
@@ -404,7 +405,7 @@ def vcf_to_zarr_sequential(
                     alleles = alleles[:n_allele]
                 elif len(alleles) < n_allele:
                     alleles = alleles + ([""] * (n_allele - len(alleles)))
-                variant_allele.append(alleles)
+                variant_alleles.append(alleles)
                 max_variant_allele_length = max(
                     max_variant_allele_length, max(len(x) for x in alleles)
                 )
@@ -422,7 +423,10 @@ def vcf_to_zarr_sequential(
 
             variant_id = np.array(variant_ids, dtype="O")
             variant_id_mask = variant_id == "."
-            variant_allele = np.array(variant_allele, dtype="O")
+            if len(variant_alleles) == 0:
+                variant_allele = np.empty((0, n_allele), dtype="O")
+            else:
+                variant_allele = np.array(variant_alleles, dtype="O")
 
             ds: xr.Dataset = create_genotype_call_dataset(
                 variant_contig_names=variant_contig_names,

--- a/sgkit/io/vcfzarr_reader.py
+++ b/sgkit/io/vcfzarr_reader.py
@@ -340,6 +340,9 @@ def _concat_zarrs_optimized(
         # copy top-level attributes
         output_zarr.attrs.update(first_zarr_group.attrs)
 
+    # consolidate metadata
+    zarr.consolidate_metadata(str(output))
+
 
 def _to_zarr(  # type: ignore[no-untyped-def]
     arr,

--- a/sgkit/tests/io/vcf/test_utils.py
+++ b/sgkit/tests/io/vcf/test_utils.py
@@ -6,7 +6,7 @@ import fsspec
 import pytest
 from callee.strings import StartsWith
 
-from sgkit.io.vcf.utils import build_url, temporary_directory
+from sgkit.io.vcf.utils import build_url, chunks, temporary_directory
 
 
 def directory_with_file_scheme() -> str:
@@ -89,3 +89,19 @@ def test_build_url():
         build_url("http://host/a%20path", "subpath") == "http://host/a%20path/subpath"
     )
     assert build_url("http://host/a path", "subpath") == "http://host/a%20path/subpath"
+
+
+@pytest.mark.parametrize(
+    "x,n,expected_values",
+    [
+        (0, 1, [[]]),
+        (1, 1, [[0]]),
+        (4, 1, [[0], [1], [2], [3]]),
+        (4, 2, [[0, 1], [2, 3]]),
+        (5, 2, [[0, 1], [2, 3], [4]]),
+        (5, 5, [[0, 1, 2, 3, 4]]),
+        (5, 6, [[0, 1, 2, 3, 4]]),
+    ],
+)
+def test_chunks(x, n, expected_values):
+    assert [list(i) for i in chunks(iter(range(x)), n)] == expected_values

--- a/sgkit/tests/io/vcf/test_vcf_reader.py
+++ b/sgkit/tests/io/vcf/test_vcf_reader.py
@@ -206,6 +206,29 @@ def test_vcf_to_zarr__parallel(shared_datadir, is_path, tmp_path):
 
 @pytest.mark.parametrize(
     "is_path",
+    [True, False],
+)
+def test_vcf_to_zarr__empty_region(shared_datadir, is_path, tmp_path):
+    path = path_for_test(shared_datadir, "CEUTrio.20.21.gatk3.4.g.vcf.bgz", is_path)
+    output = tmp_path.joinpath("vcf_concat.zarr").as_posix()
+    regions = "23"
+
+    vcf_to_zarr(path, output, regions=regions)
+    ds = xr.open_zarr(output)  # type: ignore[no-untyped-call]
+
+    assert ds["sample_id"].shape == (1,)
+    assert ds["call_genotype"].shape == (0, 1, 2)
+    assert ds["call_genotype_mask"].shape == (0, 1, 2)
+    assert ds["call_genotype_phased"].shape == (0, 1)
+    assert ds["variant_allele"].shape == (0, 4)
+    assert ds["variant_contig"].shape == (0,)
+    assert ds["variant_id"].shape == (0,)
+    assert ds["variant_id_mask"].shape == (0,)
+    assert ds["variant_position"].shape == (0,)
+
+
+@pytest.mark.parametrize(
+    "is_path",
     [False],
 )
 def test_vcf_to_zarr__parallel_temp_chunk_length(shared_datadir, is_path, tmp_path):

--- a/sgkit/tests/test_regenie.py
+++ b/sgkit/tests/test_regenie.py
@@ -243,7 +243,7 @@ def check_simulation_result(
 
     # Load simulated data
     with zarr.ZipStore(str(dataset_dir / "genotypes.zarr.zip"), mode="r") as store:
-        ds = xr.open_zarr(store)  # type: ignore[no-untyped-call]
+        ds = xr.open_zarr(store, consolidated=False)  # type: ignore[no-untyped-call]
         df_covariate = load_covariates(dataset_dir)
         df_trait = load_traits(dataset_dir)
         contigs = ds["variant_contig"].values

